bin/docserv

```diff
@@ -226,42 +260,86 @@ class Deliverable:
+            result = self.execute(commands[i], thread_id)
+            if not result: # abort if one command failed
+                return self.finish(False)
+
+            if commands[i]['post_cmd_hook'] is not None:
+                getattr(self, commands[i]['post_cmd_hook'])(commands[i])
+
+        return self.finish(result)
+
+    def execute(self, command, thread_id):
+        cmd = shlex.split(command['cmd'])
+        logger.debug("Thread %i: %s" % (thread_id, cmd))
+        s = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        self.out, self.err = s.communicate()
+
+        if command['ret_val'] is not None and not command['ret_val'] == int(s.returncode):
+            logger.warning("Thread %i: Build failed! Unexpected return value %i for '%s'" % (
+                thread_id, s.returncode, command['cmd']))
+            logger.warning("Thread %i STDOUT: %s" % (thread_id, self.out.decode('utf-8')))
+            logger.warning("Thread %i STDERR: %s" % (thread_id, self.err.decode('utf-8')))
+            return False
         return True
 
-    def parse_command(self, command, thread_id):
-        if 'replace' in command:
-            # currently only read from file logic
-            logger.debug("Reading %s" % command['replace']['filepath'])
-            f = open(command['replace']['filepath'], 'r')
-            content = str(f.read()).strip()
-            logger.debug("Content: %s" % content)
-            cmd = command['cmd'].replace(command['replace']['string'], content)
+    def finish(self, result):
+        with self.parent.deliverables_open_lock:
+            if result:
+                self.parent.deliverables[self.id]['status'] = "success"
+            else:
+                self.parent.deliverables[self.id]['status'] = "fail"
+        with self.parent.deliverables_building_lock:
+            self.parent.deliverables_building.remove(self.id)
+        if result:
+            with self.parent.deliverables_open_lock:
+                self.parent.deliverables[self.id]['successful_build_commit'] = self.parent.build_instruction['commit']
+        return result
+
+    def parse_d2d_filelist(self, command):
+        # currently only read from file logic
+        f = open(os.path.join(command['tmp_build_target'],'filelist'), 'r')
+        for line in f.readline().strip():
+            if '_bigfile.xml' not in line:
+                self.path.append(line)
+        logger.debug("Deliverable build results: %s" % str(self.path))
+        command['cmd'] = command['cmd'].replace('__FILELIST__', ' '.join(self.path))
+        return command
+
+    def extract_root_id(self, command):
+        # extract root id from DC file and then title from bigfile
+        dc_path = os.path.join(self.parent.local_repo_build_dir, self.dc_file)
+        with open(dc_path) as f:
+            import re
+            for line in f.readlines():
+                m = re.search('ROOTID=\"(.+?)\"', line)
```

Comment:
@sknorr is THIS reasonable?

Reply:
Almost. I'd hunt for the line with `^\s*ROOTID\s*=` in the first step, and then remove `^\s*ROOTID\s*=\s*["']?` and `["']?\s*` in the end.

Reply:
Thanks.